File: harness/oracle/cases/turn-extras.ts
/**
 * Oracle case (P4.D82): turn extras — the parts of an outgoing payload that are
 * not context (v4 `f933ba9c`, bug 70).
 *
 * Drives the REAL functions from v4's lib/services/chat-message/turn-extras.ts:
 * extractToolNames, buildToolChangeNotice, collectTurnExtras. The last one also
 * pulls v4's real buildAgentModeInstructions and its real token estimator, so
 * the reservation arithmetic is compared end to end rather than re-derived.
 *
 * Strings are compared EXACTLY: the tool-change notice is a contractual
 * UI-adjacent sentence the model reads, and the agent-mode instructions are a
 * prompt block.
 *
 * The provider only selects the estimator's chars-per-token rate; every row uses
 * OPENAI, whose rate is the default 3.5 the Rust port injects. (An empty plugin
 * registry — this case initializes none — is exactly what makes the default
 * apply on the v4 side too.)
 *
 * The unserializable-definition arm of countToolSchemaTokens is NOT covered here:
 * its input is a circular object, so no NDJSON row can carry it. See the
 * token-estimation case header.
 *
 * Run from inside the server checkout:
 *   cd ~/source/quilltap-server
 *   npx tsx ~/source/quilltap-v5/harness/oracle/cases/turn-extras.ts \
 *     > /tmp/oracle-turn-extras.ndjson
 */

import {
  collectTurnExtras,
  buildToolChangeNotice,
  extractToolNames,
} from '@/lib/services/chat-message/turn-extras';

const OPENAI_STYLE_TOOL = {
  type: 'function',
  function: {
    name: 'doc_read',
    description: 'Read a document from the vault by path.',
    parameters: {
      type: 'object',
      properties: {
        path: { type: 'string', description: 'Vault-relative path to the file.' },
      },
      required: ['path'],
    },
  },
};

const ANTHROPIC_STYLE_TOOL = {
  name: 'doc_write',
  description: 'Write a document into the vault.',
  input_schema: {
    type: 'object',
    properties: { path: { type: 'string' }, content: { type: 'string' } },
  },
};

const NAMELESS_TOOL = { nonsense: true };
/** An empty `function.name` is falsy, so v4's `||` chain falls to `name`. */
const EMPTY_NAME_FALLS_THROUGH = { type: 'function', function: { name: '' }, name: 'fallback_name' };
/** Both candidates falsy → 'unknown' → filtered out of the roster entirely. */
const BOTH_EMPTY = { function: { name: '' }, name: '' };
const NULL_NAME = { name: null };

type Row =
  | { kind: 'names'; id: string; tools: unknown[]; out: string[] }
  | { kind: 'notice'; id: string; toolNames: string[]; out: string }
  | {
      kind: 'extras';
      id: string;
      tools: unknown[];
      agentModeEnabled: boolean;
      agentModeMaxTurns: number;
      toolSettingsChanged: boolean;
      out: {
        agentModeInstructions: string | null;
        toolChangeNotice: string | null;
        toolSchemaTokens: number;
        reservedTokens: number;
      };
    };

const rows: Row[] = [];

// extractToolNames(tools)
const nameCases: Array<[string, unknown[]]> = [
  ['empty', []],
  ['both-shapes', [OPENAI_STYLE_TOOL, ANTHROPIC_STYLE_TOOL]],
  ['drops-unrecognisable', [OPENAI_STYLE_TOOL, ANTHROPIC_STYLE_TOOL, NAMELESS_TOOL]],
  ['empty-function-name-falls-through', [EMPTY_NAME_FALLS_THROUGH]],
  ['both-empty-dropped', [BOTH_EMPTY]],
  ['null-name-dropped', [NULL_NAME]],
  ['order-preserved', [ANTHROPIC_STYLE_TOOL, OPENAI_STYLE_TOOL]],
];
for (const [id, tools] of nameCases) {
  rows.push({ kind: 'names', id, tools, out: extractToolNames(tools) });
}

// buildToolChangeNotice(toolNames)
const noticeCases: Array<[string, string[]]> = [
  ['none-left', []],
  ['one', ['doc_read']],
  ['two', ['doc_read', 'doc_write']],
  ['many', ['a', 'b', 'c', 'd', 'e']],
];
for (const [id, toolNames] of noticeCases) {
  rows.push({ kind: 'notice', id, toolNames, out: buildToolChangeNotice(toolNames) });
}

// collectTurnExtras(options)
const extrasCases: Array<[string, unknown[], boolean, number, boolean]> = [
  // tools, agent mode off, no change → the schemas are the whole reservation
  ['schemas-only', [OPENAI_STYLE_TOOL], false, 25, false],
  // nothing at all → nothing reserved
  ['nothing', [], false, 25, false],
  // agent mode alone (no tools) — the instructions still cost a system message
  ['agent-mode-no-tools', [], true, 25, false],
  ['agent-mode-with-tools', [OPENAI_STYLE_TOOL], true, 25, false],
  // the maxTurns figure is interpolated into the instructions, so it moves the count
  ['agent-mode-100-turns', [OPENAI_STYLE_TOOL], true, 100, false],
  // tool-change notice alone, naming the roster
  ['tool-change-with-tools', [OPENAI_STYLE_TOOL, ANTHROPIC_STYLE_TOOL], false, 25, true],
  // the "all disabled" notice fires on an EMPTY slate
  ['tool-change-all-disabled', [], false, 25, true],
  // both injections accumulate
  ['both-injections', [OPENAI_STYLE_TOOL, ANTHROPIC_STYLE_TOOL], true, 25, true],
  // an unrecognisable entry is still counted in the schemas but never named
  ['unnamed-tool-still-measured', [NAMELESS_TOOL], false, 25, true],
];
for (const [id, tools, agentModeEnabled, agentModeMaxTurns, toolSettingsChanged] of extrasCases) {
  rows.push({
    kind: 'extras',
    id,
    tools,
    agentModeEnabled,
    agentModeMaxTurns,
    toolSettingsChanged,
    out: collectTurnExtras({
      tools,
      agentMode: { enabled: agentModeEnabled, maxTurns: agentModeMaxTurns },
      toolSettingsChanged,
      provider: 'OPENAI' as never,
    }),
  });
}

for (const r of rows) process.stdout.write(JSON.stringify(r) + '\n');

File: crates/quilltap-core/src/services/turn_extras.rs
//! Turn Extras — the parts of an outgoing payload that are not context.
//!
//! Port of v4's `lib/services/chat-message/turn-extras.ts` (`f933ba9c`, bug 70).
//!
//! The context builder produces the messages; the orchestrator then adds things
//! the builder never sees: the tool/function schemas (which ride beside the
//! message array entirely) and one or two system messages spliced in downstream
//! (agent-mode instructions, the tool-change notice).
//!
//! All of it counts against the model's context window, and none of it used to
//! count against the budget — the builder packed history right up to the ceiling
//! and the orchestrator then piled these on top, which is how a turn could clear
//! the budget and still exceed the window.
//!
//! **This module is the one place those additions are built and measured.**
//! Building them here and handing the *same strings* back to the splice sites is
//! the point: a reservation computed from separately-constructed text is a
//! reservation that drifts the first time either copy is edited.
//!
//! Seam: v4 passes the `Provider` to the estimator, which resolves its
//! chars-per-token from the plugin registry. v5 injects the rate instead (the
//! established [`crate::token_estimation`] seam), so `chars_per_token` is a
//! parameter here.

use serde_json::Value;

use crate::token_estimation::{count_message_tokens, count_tool_schema_tokens};

/// Extract tool names from provider-shaped tool definitions — v4
/// `extractToolNames`.
///
/// Handles both the OpenAI-style `{type, function: {name}}` wrapper and the flat
/// `{name}` shape; anything unrecognisable is dropped rather than named (v4 maps
/// it to the literal `'unknown'` and then filters that out).
///
/// v4's `toolObj.function?.name || toolObj.name` is a truthiness chain, so an
/// EMPTY name falls through to the next candidate exactly as an absent one does.
/// A non-string name is out of the modeled domain: every producer of this slate
/// (v4's tool builders and v5's [`crate::tools`]) types the name as a string.
/// (Likewise a `null`/primitive ARRAY ENTRY: v4's `toolObj.function?.name`
/// optional-chains only past `.function`, so a null entry throws there where
/// this walk drops it — out of the modeled domain, noted not modeled.)
pub fn extract_tool_names(tools: &[Value]) -> Vec<String> {
    tools
        .iter()
        .filter_map(|tool| {
            let truthy = |v: Option<&Value>| {
                v.and_then(Value::as_str)
                    .filter(|s| !s.is_empty())
                    .map(str::to_string)
            };
            truthy(tool.get("function").and_then(|f| f.get("name")))
                .or_else(|| truthy(tool.get("name")))
        })
        .collect()
}

/// Build the notice telling the model its tool roster changed mid-chat — v4
/// `buildToolChangeNotice`.
///
/// Only fires when the user explicitly changed tool settings — not on the first
/// message and not on an interval.
pub fn build_tool_change_notice(tool_names: &[String]) -> String {
    if tool_names.is_empty() {
        return "[System Notice] Your available tools have been updated. All tools have been disabled for this chat. Do not attempt to use any tools.".to_string();
    }

    format!(
        "[System Notice] Your available tools have been updated. You now have access to the following {} tool(s): {}. Tools not listed are no longer available for this chat.",
        tool_names.len(),
        tool_names.join(", ")
    )
}

/// v4 `TurnExtrasOptions`.
pub struct TurnExtrasOptions<'a> {
    /// Tools as they will be sent (already narrowed to the native-calling set).
    pub tools: &'a [Value],
    /// Whether agent mode is on for this turn.
    pub agent_mode_enabled: bool,
    /// The agent-mode turn ceiling (only read when enabled).
    pub agent_mode_max_turns: i64,
    /// Whether the user changed tool settings since the last turn.
    pub tool_settings_changed: bool,
    /// The estimator's chars-per-token rate (v4 passes the provider instead).
    pub chars_per_token: f64,
}

/// v4 `TurnExtras`.
#[derive(Clone, Debug, PartialEq)]
pub struct TurnExtras {
    /// Agent-mode instructions to splice in, or `None` when agent mode is off.
    pub agent_mode_instructions: Option<String>,
    /// Tool-change notice to splice in, or `None` when tools did not change.
    pub tool_change_notice: Option<String>,
    /// Tokens the tool schemas occupy beside the message array.
    pub tool_schema_tokens: i64,
    /// Total tokens these additions will contribute to the outgoing payload —
    /// pass as `reserved_outgoing_tokens` to the context builder so history is
    /// not packed into space they will take.
    pub reserved_tokens: i64,
}

/// Build and measure everything this turn will add after the context is built —
/// v4 `collectTurnExtras`.
///
/// Call before building context (all the inputs are known by then), pass
/// `reserved_tokens` into the builder, and splice the returned strings rather
/// than rebuilding them.
pub fn collect_turn_extras(options: TurnExtrasOptions<'_>) -> TurnExtras {
    let agent_mode_instructions = if options.agent_mode_enabled {
        Some(super::agent_mode::build_agent_mode_instructions(
            options.agent_mode_max_turns,
        ))
    } else {
        None
    };

    let tool_change_notice = if options.tool_settings_changed {
        Some(build_tool_change_notice(&extract_tool_names(options.tools)))
    } else {
        None
    };

    let tool_schema_tokens = count_tool_schema_tokens(options.tools, options.chars_per_token);

    // Both extras are spliced in as system messages, so they carry the same
    // per-message overhead the builder's own accounting applies.
    let agent_mode_tokens = agent_mode_instructions
        .as_deref()
        .map(|s| count_message_tokens("system", s, options.chars_per_token))
        .unwrap_or(0);
    let tool_change_tokens = tool_change_notice
        .as_deref()
        .map(|s| count_message_tokens("system", s, options.chars_per_token))
        .unwrap_or(0);

    TurnExtras {
        agent_mode_instructions,
        tool_change_notice,
        tool_schema_tokens,
        reserved_tokens: tool_schema_tokens + agent_mode_tokens + tool_change_tokens,
    }
}
